### PR TITLE
Round relative time interval arguments

### PR DESCRIPTION
**Pull request by bogdan-szabo-sociomantic**

### CODE REVIEW

File: relnotes/timedivision.feature.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+### Round relative time interval arguments
+
+`ocean.text.arguments.TimeIntervalArgs`
+
+Relative values (eg. 1m, 4h) and 'now' can be rounded by [m]inutes, [h]ours and [d]ays.
```
Comment (christoph-wasicki-sociomantic) [inline relnotes/timedivision.feature.md:7]: "rounded" here means actually floored, right? E.g. `-t 1h/h 1h` at 15:59` would also start at `14:00`.

File: relnotes/timedivision.feature.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+
+      eg. -t 1h/h 1h = if the command is executed at 15:23:23 then
```
Comment (christoph-wasicki-sociomantic) [inline relnotes/timedivision.feature.md:10]: Minor: I did not know the format of time intervals and was confused before @bogdan-szabo-sociomantic explained it to me. The doc could repeat the meaning here `<time_start>/<floor_to_time_unit> [time_end]` or use an example that's less ambiguous (e.g. `4h/h 2h`).

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 3]
```diff
@@ -127,6 +143,50 @@ public istring validateTimeIntervalArgs ( Arguments args )
     return null;
 }
 
+/******************************************************************************
+
+    Parse one time interval argument
+
+    Params:
+        op = the math operation performed tot the reference time.
+             Only "+", "-" is permitted.
```
Comment (liza-churkyn-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d:153]: `tot` => `to`

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 4]
```diff
@@ -388,6 +434,71 @@ unittest
...
+
+    setupTimeIntervalArgs(args, false);
+    args.parse("--time-interval now/h");
+    interval = processTimeIntervalArgs(args);
+
+    timeToUnixTime("2019-04-01 15:00:00", dummy_time, dummy_conv);
+    Test.test!("==")(interval.begin, dummy_time);
+
+    timeToUnixTime("2019-04-01 15:26:31", dummy_time, dummy_conv);
+    Test.test!("==")(interval.end, dummy_time);
```
Comment (liza-churkyn-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d:554]: you have both tests for hour and minutes, one test with a `/d` (day) would also be good

File: relnotes/timedivision.feature.md (hunk 1 above)
Comment (bogdan-szabo-sociomantic) [reply relnotes/timedivision.feature.md:7]: `-t 1h 1h` at `15:59` means `14:59 to 15:59` 

`-t 1h/h 1h` at `15:59` means `14:00 to 15:00` 

so, yes

File: src/ocean/text/arguments/TimeIntervalArgs.d (hunk 4 above)
Comment (bogdan-szabo-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d:554]: done

File: relnotes/timedivision.feature.md (hunk 2 above)
Comment (bogdan-szabo-sociomantic) [reply relnotes/timedivision.feature.md:10]: done

**Comment by bogdan-szabo-sociomantic** [comment]
ping @matthias-wende-sociomantic

**Review by christoph-wasicki-sociomantic — APPROVED** [review]

File: relnotes/timedivision.feature.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
+### Round relative time interval arguments
+
+`ocean.text.arguments.TimeIntervalArgs`
+
+Handler for time interval CLI arguments.
+
+Relative values (eg. 1m, 4h) and 'now' can be rounded by [m]inutes, [h]ours and [d]ays.
+The new relative format is [time_start][time unit]/[floor_to_time_unit] or now/[floor_to_time_unit].
+
+      eg. -t 1h/h 1h = if the command is executed at 15:23:23 then
```
Comment (matthias-wende-sociomantic) [inline relnotes/timedivision.feature.md:10]: The example is different to what the description says.

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 6]
```diff
@@ -86,12 +100,14 @@ public void setupTimeIntervalArgs ( Arguments args, bool required )
...
-        .help("Specify a time interval. It 1 or 2 values, where they can be:\n\t" ~
+        .help("Specify a time interval. It takes 1 or 2 values, where they can be:\n\t" ~
             "'now' an alias for the current timestamp\n\t" ~
             "'yesterday' an alias for yesterdays date\n\t" ~
             "an integer for unix timestamps\n\t" ~
             "a time duration `{int}m`. Supported units are [m]inutes, [h]ours and [d]ays\n\t" ~
-            "an iso1806 date (YYYY-MM-DD)");
+            "an iso1806 date (YYYY-MM-DD)\n\t\n\t"~
+            "the relative values and 'now' can be rounded by [m]inutes, [h]ours and [d]ays.\n\t" ~
+            "eg. 1h/h will set the minutes and seconds to `0`");
```
Comment (matthias-wende-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d:110]: I think the help should fully document how to use the round feature.

**Review by matthias-wende-sociomantic — COMMENTED** [review]
First review. Two minor documentation comments and could you please provide a description. I'll review the code soon.

File: relnotes/timedivision.feature.md (hunk 5 above)
Comment (bogdan-szabo-sociomantic) [reply relnotes/timedivision.feature.md:10]: how is it different?

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 7]
```diff
@@ -457,54 +581,263 @@ unittest
...
-        true, if the string is a valid interval
+        true, if the string is a valid relative time
 
 ******************************************************************************/
 
-public bool isTimeInterval ( cstring value )
+public bool isRelativeTime ( cstring value )
 {
-    if (value == "")
+    if ( value == "" )
```
Comment (matthias-wende-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d]: This is a stylistic change and is not related to the commit.

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 8]
```diff
@@ -457,54 +581,263 @@ unittest
...
+
+    Params:
+        value = the string time
+
+    Returns:
+        the char representing the time unit or '?' in case of error
+
+******************************************************************************/
+
+char getTimeUnit ( cstring value )
```
Comment (matthias-wende-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d]: missing protection attribute

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 9]
```diff
@@ -457,54 +581,263 @@ unittest
 
 /******************************************************************************
 
-    Checks if a string can be converted to a time interval
+    Checks if a string can be converted to a time stamp
```
Comment (matthias-wende-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d]: That's not a very accurate definition of the function. What about `Checks if a string is a valid relative time` and maybe add an example.

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 10]
```diff
@@ -457,54 +581,263 @@ unittest
...
+    return isValidTimeUnit(value[value.length - 1]);
+}
+
+/// Validations for relative times
+unittest
+{
+    Test.test!("==")(isRelativeTime(""), false);
+    Test.test!("==")(isRelativeTime("1s"), true);
+    Test.test!("==")(isRelativeTime("1m"), true);
+    Test.test!("==")(isRelativeTime("2m"), true);
```
Comment (matthias-wende-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d:663]: This test seems to indirectly test the `toLong` function but doesn't add any valuable new test to `isRelativeTime`.

Comment (matthias-wende-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d:663]: actually that's true for most of the tests here that either test `toLong` or `isValidTimeUnit` indirectly.

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 11]
```diff
@@ -457,54 +581,263 @@ unittest
...
+
+    Params:
+        value = the relative time that will be validated
+
+    Returns:
+        true, if the string matches the pattern [some value]/[time unit]
+
+******************************************************************************/
+
+public bool isDividedTime ( cstring value )
```
Comment (matthias-wende-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d]: why is this public?

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 12]
```diff
@@ -457,54 +581,263 @@ unittest
...
+            unit. The supported units are seconds, minutes, hours and days.
 
     Returns:
-        true, if the string is a valid interval
+        true, if the string is a valid relative time
 
 ******************************************************************************/
 
-public bool isTimeInterval ( cstring value )
+public bool isRelativeTime ( cstring value )
```
Comment (matthias-wende-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d]: Why is this a public function?

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 13]
```diff
@@ -457,54 +581,263 @@ unittest
...
+
+    Params:
+        unit = the char to be checked
+
+    Returns:
+        true if the unit is valid
+
+******************************************************************************/
+
+public bool isValidTimeUnit ( char unit )
```
Comment (matthias-wende-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d]: public?

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 14]
```diff
@@ -457,54 +581,263 @@ unittest
...
+
+    Params:
+        value = the relative time
+
+    Returns:
+        the relative time without the division
+
+******************************************************************************/
+
+public cstring removeTimeDivision ( cstring value )
```
Comment (matthias-wende-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d]: Again why is this a public method?

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 15]
```diff
@@ -457,54 +581,263 @@ unittest
...
+    Params:
+        value = the timestamp
+        unit = the time unit used for rounding
+
+    Returns:
+        the rounded timestamp
+
+******************************************************************************/
+
+long roundTime ( long value, char unit )
```
Comment (matthias-wende-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d]: Missing protection attribute.

**Review by matthias-wende-sociomantic — COMMENTED** [review]
This PR seems to do more then what it says in the title. You might want to explain why you refactored the module and may split the changes into better structured smaller commits.

Also it seems that exposing all methods as public is not necessary.

**Comment by bogdan-szabo-sociomantic** [comment]
@ben-palmer-sociomantic can you have a look at this pr?

File: src/ocean/text/arguments/TimeIntervalArgs.d (hunk 9 above)
Comment (bogdan-szabo-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d]: I think there are enough examples in the unittests

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 16]
```diff
@@ -127,6 +127,46 @@ public istring validateTimeIntervalArgs ( Arguments args )
...
+******************************************************************************/
+
+public long parseTimeArgument ( cstring op ) ( cstring value, long reference_time,
+    bool include_end_date )
+{
+    long result;
+
+    if ( isTimeInterval(value) )
+    {
+        static if ( op == "+" )
```
Comment (ben-palmer-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d:168]: Maybe add a static assert on the value of `op`?

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 17]
```diff
@@ -142,48 +182,33 @@ public TimestampInterval processTimeIntervalArgs ( Arguments args )
     enforce(num_args >= 1 && num_args <= 2, "Not enough arguments provided");
     bool include_end_date = !args["time-interval-exclude"].set;
 
-    long begin, end;
```
Comment (ben-palmer-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d:145]: Should the docs for this method be updated as the functionality has changed?

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 18]
```diff
@@ -517,7 +517,7 @@ unittest
 
 ******************************************************************************/
 
-public bool isTimeInterval ( cstring value )
+private bool isRelativeTime ( cstring value )
```
Comment (ben-palmer-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d:635]: This is a change to the public API that would need to be in a major release.

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 19]
```diff
@@ -86,12 +100,14 @@ public void setupTimeIntervalArgs ( Arguments args, bool required )
     auto interval = args("time-interval")
         .aliased('t')
         .params(1, 2)
-        .help("Specify a time interval. It 1 or 2 values, where they can be:\n\t" ~
+        .help("Specify a time interval. It takes 1 or 2 values, where they can be:\n\t" ~
```
Comment (ben-palmer-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d:103]: This documentation change is not related to this commit.

**Review by ben-palmer-sociomantic — COMMENTED** [review]
I had a couple of comments.

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 20]
```diff
@@ -356,7 +382,7 @@ unittest
     interval = processTimeIntervalArgs(args);
 
     Test.test!("==")(interval.begin, TEST_TIME_NOW - 60);
-    Test.test!("==")(interval.end, TEST_TIME_NOW + 59);
+    Test.test!("==")(interval.end, TEST_TIME_NOW - 1);
```
Comment (liza-churkyn-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d:445]: This commit contains two changes: 1. moving the duplicate part out 2. functionality change, where in case of 2 arguments `end` becomes relating to `begin` and not relative to `now`. Due to the functionality change you had to update the unit test.  

So, you could split this commit to two: the 1st commit would contain the unit test update `Test.test!("==")(interval.end, TEST_TIME_NOW - 1);` and the functionality change:
```
    begin = time(null) - parseTimeInterval(begin_arg);	
    end = begin + parseTimeInterval(end_arg);
```

Instead of the previous:
```
    begin = time(null) - parseTimeInterval(begin_arg);	
    end = time(null) + parseTimeInterval(end_arg);
```

The 2nd commit can contain the rest.

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 21]
```diff
@@ -200,7 +220,7 @@ public TimestampInterval processTimeIntervalArgs ( Arguments args )
 
     return createTimestampInterval(
         args["time-interval"].assigned[0],
-        args["time-interval"].assigned[0],
+        removeTimeDivision(args["time-interval"].assigned[0]),
```
Comment (liza-churkyn-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d:247]: Why are you calling `removeTimeDivision` here? Here end = begin = "non relative time"...

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 22]
```diff
@@ -226,7 +246,7 @@ private TimestampInterval createTimestampInterval ( cstring str_begin,
 
     if ( !isRelativeTime(str_end) )
     {
-        reference_time = parseDateString(str_end, include_end_date);
+        reference_time = parseDateString(removeTimeDivision(str_end), include_end_date);
```
Comment (liza-churkyn-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d:273]: Same question as above, don't we remove division only from "relative time" values?

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 23]
```diff
@@ -553,7 +645,222 @@ unittest
...
+    Params:
+        unit = the char to be checked
+
+    Returns:
+        true if the unit is valid
+
+******************************************************************************/
+
+private bool isValidTimeUnit ( char unit )
+{
```
Comment (liza-churkyn-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d:687]: Taking out the `unit == 's' || unit == 'm' || unit == 'h' || unit == 'd'` code and putting it a separate function `isValidTimeUnit` can be in its own commit (before this commit)

File: src/ocean/text/arguments/TimeIntervalArgs.d (hunk 18 above)
Comment (bogdan-szabo-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d:635]: I don't think that there is anyone who uses this function...

File: src/ocean/text/arguments/TimeIntervalArgs.d (hunk 17 above)
Comment (bogdan-szabo-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d:145]: I think it's better to update the release notes

File: src/ocean/text/arguments/TimeIntervalArgs.d (hunk 23 above)
Comment (bogdan-szabo-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d:687]: Should I go in that many details with the commits?

File: src/ocean/text/arguments/TimeIntervalArgs.d (hunk 6 above)
Comment (bogdan-szabo-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d:110]: I'm not sure how many details should we add here... If there is too much information, people will might just not read it

Comment (tiyash-basu-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d:110]: From https://github.com/sociomantic-tsunami/ocean/pull/712/files#r287270639, I'd recommend changing `rounded` to `floored` here and everywhere else in this context.

I agree that the flooring feature needs to be documented. The syntax is already there, perhaps how to use the syntax should be added.

The more I look at it, the more I think that the rounding unit should be a separate command line argument of its own. Squashing two different inputs into one argument should be avoided as much as possible.

File: src/ocean/text/arguments/TimeIntervalArgs.d (hunk 21 above)
Comment (bogdan-szabo-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d:247]: it's needed because of this case: `--time-interval now/h` which should be the same as `--time-interval now/h now`

File: src/ocean/text/arguments/TimeIntervalArgs.d (hunk 22 above)
Comment (bogdan-szabo-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d:273]: it's needed because `parseDateString` does not know how to parse strings with time divisions. 

`"--time-interval now/h now/m"` is a valid case

File: src/ocean/text/arguments/TimeIntervalArgs.d (hunk 18 above)
Comment (ben-palmer-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d:635]: This is an open source library so we can't actually say that for sure.

Comment (gavin-norman-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d:635]: Correct. A change like this must always be considered a breaking change. If the function must be renamed, consider using a deprecation.

Comment (bogdan-szabo-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d:635]: done

File: src/ocean/text/arguments/TimeIntervalArgs.d (hunk 6 above)
Comment (bogdan-szabo-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d:110]: we discussed with @scott-gibson-sociomantic and @liza-churkyn-sociomantic and this feature was designed to work like grafana intervals

**Comment by gavin-norman-sociomantic** [comment]
(Accidentally clicked on CLOSE, sorry!)

**Comment by gavin-norman-sociomantic** [comment]
@bogdan-szabo-sociomantic this contains a breaking change now, right? If so, it'll need a `x.breaking.md` release note please.

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 24]
```diff
@@ -127,9 +143,72 @@ public istring validateTimeIntervalArgs ( Arguments args )
...
+        {
+            result = reference_time + parseTimeInterval(removeTimeDivision(value));
+        }
+        else static if ( op == "-" )
+        {
+            result = reference_time - parseTimeInterval(removeTimeDivision(value));
+        }
+        else
+        {
+            static assert(true, "Only `+` and `-` operations are supported.");
```
Comment (ben-palmer-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d]: Should this be `false`?

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 25]
```diff
@@ -403,7 +567,7 @@ unittest
 
 ******************************************************************************/
 
-public long parseTimeInterval ( cstring value )
+private long parseTimeInterval ( cstring value )
```
Comment (ben-palmer-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d]: This is another breaking change in the API that needs to be addressed.

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 26]
```diff
@@ -457,54 +621,266 @@ unittest
...
+******************************************************************************/
+
+private cstring removeTimeDivision ( cstring value )
+{
+    if ( !isDividedTime(value) )
+    {
+        return value;
+    }
+
+    return value[0..$-2];
```
Comment (ben-palmer-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d:779]: Should we do a check on array length here?

**Review by ben-palmer-sociomantic — COMMENTED** [review]
Couple more comments from me.

File: src/ocean/text/arguments/TimeIntervalArgs.d (hunk 26 above)
Comment (bogdan-szabo-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d:779]: not really, but I added one.

File: src/ocean/text/arguments/TimeIntervalArgs.d (hunk 25 above)
Comment (bogdan-szabo-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d]: reverted to public

File: src/ocean/text/arguments/TimeIntervalArgs.d (hunk 24 above)
Comment (bogdan-szabo-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d]: yes, thanks

File: relnotes/timedivision.breaking.md [hunk 27]
```diff
@@ -0,0 +1,5 @@
+### Round relative time interval arguments
+
+`ocean.text.arguments.TimeIntervalArgs`
+
+isTimeInterval ( cstring value ) is now deprecated
```
Comment (gavin-norman-sociomantic) [inline relnotes/timedivision.breaking.md]: A method being deprecated itsn't a breaking change.

**Comment by gavin-norman-sociomantic** [comment]
@bogdan-szabo-sociomantic when you're back, let's look at this PR. I can't believe it's impossible to implement this without a breaking change. Breaking changes should be an absolute last resort. Let's find a way to merge this into v4.x.x.

File: relnotes/timedivision.breaking.md (hunk 27 above)
Comment (ben-palmer-sociomantic) [reply relnotes/timedivision.breaking.md]: Agreed. Should be `timedivision.deprecation.md`.

**Review by ben-palmer-sociomantic — COMMENTED** [review]
Tests seem to be failing now.

File: relnotes/timedivision.breaking.md (hunk 27 above)
Comment (bogdan-szabo-sociomantic) [reply relnotes/timedivision.breaking.md]: done :)

**Comment by bogdan-szabo-sociomantic** [comment]
@ben-palmer-sociomantic This started to happen after I rebased to v5... it looks like the prometheus module has a compiling error: 

```
./src/ocean/util/prometheus/collector/StatFormatter.d(216): Error: template instance ocean.util.prometheus.collector.StatFormatter.getSanitized!float error instantiating
```

**Comment by tiyash-basu-sociomantic** [comment]
@bogdan-szabo-sociomantic The prometheus module is D1, and it was added by neptune to the D2-only v5.x.x branch without applying d1to2fix on it. This is causing the error that you saw. I'd recommend waiting until its fixed (https://github.com/sociomantic-tsunami/ocean/pull/715).

**Comment by gavin-norman-sociomantic** [comment]
#715 is fixed now. But this PR should be based on v4x.x. As I said before, there's no reason it needs to be a breaking change.

**Comment by gavin-norman-sociomantic** [comment]
#715 is fixed now. But this PR should be based on v4x.x. As I said before, there's no reason it needs to be a breaking change.

**Comment by bogdan-szabo-sociomantic** [comment]
@gavin-norman-sociomantic I rebased to 4.x.x 

I am not sure if there are any breaking changes left. All the new functions are private, and for `isTimeInterval` that I renamed to `isRelativeTime`, there is an alias with the old name and marked as deprecated

**Comment by ben-palmer-sociomantic** [comment]
This doesn't seem to be running the tests for some reason.

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 28]
```diff
@@ -561,18 +561,21 @@ public bool isTimeInterval ( cstring value )
     return true;
 }
 
+/// ditto
+public deprecated alias isRelativeTime isTimeInterval;
```
Comment (gavin-norman-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d:655]: Deprecation message?

File: relnotes/timedivision.deprecation.md [hunk 29]
```diff
@@ -0,0 +1,5 @@
+### Round relative time interval arguments
```
Comment (gavin-norman-sociomantic) [inline relnotes/timedivision.deprecation.md:1]: The commit message for "Rename isTimeInterval to isRelativeTime" should mention the reason for this change. (Currently, while the code change looks fine, there's no context as to why we might want to merge this change.)

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 30]
```diff
@@ -225,15 +225,7 @@ public TimestampInterval processTimeIntervalArgs ( Arguments args )
...
-
-    if ( isTimeInterval(begin_arg) )
-    {
-        end = parseTimeArgument!("+")(end_arg, time(null), include_end_date);
-    }
-    else
-    {
-        end = parseTimeArgument!("+")(end_arg, begin, include_end_date);
-    }
+    end = parseTimeArgument!("+")(end_arg, begin, include_end_date);
```
Comment (gavin-norman-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d]: The commit message "Make end argument to be relative to the begin argument instead of the current timestamp" should describe the rationale for the change.

File: src/ocean/text/arguments/TimeIntervalArgs.d [hunk 31]
```diff
@@ -198,9 +198,10 @@ public long parseTimeArgument ( cstring op ) ( cstring value, long reference_tim
     an iso8601 date, a timestamp, `now` or `yesterday`.
 
     If only one argument is provided, the second one will be infered:
+        - `yesterday` it's rewriten as `1d/d 1d`
```
Comment (gavin-norman-sociomantic) [inline src/ocean/text/arguments/TimeIntervalArgs.d:201]: The commit "Implement yesterday keyword as relative time" appears to be changing some previous behaviour, rather than implementing a new behaviour. The commit message should describe what the old behaviour was and the reason for changing it.

Comment (gavin-norman-sociomantic) [reply src/ocean/text/arguments/TimeIntervalArgs.d:201]: It sounds like this could potentially be a breaking change, but without under understanding the old behaviour and the rationale behind changing it, it's difficult to say.

If it does turn out to be a breaking change, we could introduce a new argument, in addition to `"yesterday"`.

**Review by gavin-norman-sociomantic — COMMENTED** [review]
Partial review. General feedback: this PR is difficult to review, for the following reasons:
1. It mixes a lot of seemingly unrelated changes. Or are they related? This conjures a feeling of unease for the reviewer.
2. There's virtually no documentation of the reason for the changes. Without understanding why a change is proposed, it's very difficult to make a judgment on it.

Suggestion:
1. Split the PR into several smaller ones, each consisting of purely strictly related changes.
2. As mentioned in my review comments, enhance the commit messages to explain what the old behaviour was, what's changing, and why the changes are being made.
